cluster/kubernetes/resourcekinds.go

```diff
@@ -120,27 +122,50 @@ func (dk *deploymentKind) getPodControllers(c *Cluster, namespace string) ([]pod
 	return podControllers, nil
 }
 
+func deploymentErrors(d *apiapps.Deployment) []string {
+	var errs []string
+	for _, cond := range d.Status.Conditions {
```

Comment:
I didn't know much about conditions so I went to read the internet.

My understanding is that each entry in `.Conditions` represents a transition -- and though the Kubernetes authors seem to experience physical pain at the suggestion of this representing a state machine, at least some of the conditions represent state transitions. https://github.com/kubernetes/kubernetes/issues/7856 makes very interesting reading.

Ultimately we want to know "will this rollout succeed?" with possible answers "eventually", "definitely not", and "it already has". I don't know enough to judge whether the result compiled in this loop answers that question, but I suspect that it might end up including error messages that have been superseded. For example, it could pick up a condition saying the deployment wasn't available because of some transient situation which later went away, allowing progress.

If I'm near the mark, we need to take some more care here to interpret the conditions -- e.g., if the most recent condition says "everything's OK" we can ignore the rest. I could be wrong, and just not understanding what you've already thought through -- in which case, I'd appreciate a comment above the code explaining what it's doing.

Reply:
> it might end up including error messages that have been superseded. For example, it could pick up a condition saying the deployment wasn't available because of some transient situation which later went away, allowing progress.

I think if error went away, appropriate condition changed as well.
For example, if deployment wasn't available (`cond.Type == apiapps.DeploymentAvailable && cond.Status == apiv1.ConditionFalse`) but now it is available, then this condition is not `False` anymore:
`cond.Type == apiapps.DeploymentAvailable && cond.Status != apiv1.ConditionFalse`

Reply:
I was quite surprised to see that conditions are totally ignored (i.e., never populated) for everything except for Deployments!